src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -314,6 +344,49 @@ def classpath_dest_filename(coord: str, src_filename: str) -> str:
     return f"{dest_name}{ext}"
 
 
+@dataclass(frozen=True)
+class ArtifactRequirementsWithLocalFiles:
+    # omg tidy up this name. it's bad
+    artifact_requirements: ArtifactRequirements
+    digest: Digest
+
+
+@rule
+async def use_local_artifacts_where_possible(
```

Comment:
It should be pretty easy to make this support `DownloadFile` for handling the `url` field. Note that the cache benefits would be realised infrequently since the download is only done when resolving the lockfile, not when building.